File: api/src/services/emissionsService.ts
import {
  AnalyzeTransactionsRequestSchema,
  DEFAULT_TRANSACTION_LIMIT,
  type AnalyzeTransactionsRequest,
  type AnalyzeTransactionsResponse,
  type AnalyzedTransaction,
} from "@carboniq/contracts";
import {
  CATEGORY_ORDER,
  EMISSION_FACTORS,
  KEYWORD_CATEGORY_MAP,
  MCC_CATEGORY_MAP,
  REFERENCE_ANALYZED_AT,
  type AnalysisSnapshot,
  type EmissionCategoryValue,
  type RawTransaction,
} from "../lib/aiRules.js";
import { roundTo } from "../lib/aiMath.js";
import { spendEmissionsProvider } from "./spendEmissionsProvider.js";
import { transactionProvider } from "./transactionProviderService.js";

function emptyCategoryTotals(): Record<EmissionCategoryValue, number> {
  return Object.fromEntries(CATEGORY_ORDER.map((category) => [category, 0])) as Record<
    EmissionCategoryValue,
    number
  >;
}

class EmissionsService {
  private readonly analysisCache = new Map<string, AnalyzeTransactionsResponse>();

  private getCacheKey(wallet: string, limit: number, plaidAccessToken?: string) {
    return `${wallet}:${limit}:${plaidAccessToken ?? ""}`;
  }

  classifyTransaction(rawTransaction: RawTransaction): EmissionCategoryValue {
    if (rawTransaction.mccCode && MCC_CATEGORY_MAP[rawTransaction.mccCode]) {
      return MCC_CATEGORY_MAP[rawTransaction.mccCode];
    }

    const lowered = rawTransaction.description.toLowerCase();
    for (const [keywords, category] of KEYWORD_CATEGORY_MAP) {
      if (keywords.some((keyword) => lowered.includes(keyword))) {
        return category;
      }
    }

    return "other";
  }

  private async buildSnapshot(
    wallet: string,
    limit: number,
    plaidAccessToken?: string
  ): Promise<
    AnalysisSnapshot & {
      source: "upload" | "seeded";
      sourceLabel?: string;
    }
  > {
    const ledger = await transactionProvider.getTransactions(wallet, plaidAccessToken);
    const selected = ledger.transactions.slice(0, limit);

    const categoryBreakdown = emptyCategoryTotals();
    const categorySpend = emptyCategoryTotals();
    const transactions: AnalyzedTransaction[] = [];
    let totalSpendUsd = 0;
    let totalCo2eGrams = 0;

    for (const rawTransaction of selected) {
      const category = this.classifyTransaction(rawTransaction);
      const providerEstimate = await spendEmissionsProvider.estimate(rawTransaction);
      const emissionFactor =
        providerEstimate?.emissionFactor ?? EMISSION_FACTORS[category];
      const co2eGrams =
        providerEstimate?.co2eGrams ??
        roundTo(rawTransaction.amountUsd * emissionFactor, 2);

      transactions.push({
        transactionId: rawTransaction.transactionId,
        description: rawTransaction.description,
        amountUsd: rawTransaction.amountUsd,
        mccCode: rawTransaction.mccCode,
        category,
        co2eGrams,
        emissionFactor: roundTo(emissionFactor, 2),
        date: rawTransaction.date,
      });

      totalSpendUsd = roundTo(totalSpendUsd + rawTransaction.amountUsd, 2);
      totalCo2eGrams = roundTo(totalCo2eGrams + co2eGrams, 2);
      categorySpend[category] = roundTo(
        categorySpend[category] + rawTransaction.amountUsd,
        2
      );
      categoryBreakdown[category] = roundTo(
        categoryBreakdown[category] + co2eGrams,
        2
      );
    }

    const response = {
      wallet,
      transactionCount: transactions.length,
      totalCo2eGrams,
      categoryBreakdown,
      transactions,
      analyzedAt: REFERENCE_ANALYZED_AT,
    } satisfies AnalyzeTransactionsResponse;

    return {
      response,
      totalSpendUsd,
      totalCo2eGrams,
      categorySpendTotals: categorySpend,
      categoryEmissionTotals: categoryBreakdown,
      source: ledger.source,
      sourceLabel: ledger.sourceLabel,
    };
  }

  async analyzeTransactions(
    rawRequest: AnalyzeTransactionsRequest
  ): Promise<AnalyzeTransactionsResponse> {
    const request = AnalyzeTransactionsRequestSchema.parse(rawRequest);
    const cacheKey = this.getCacheKey(
      request.wallet,
      request.limit,
      request.plaidAccessToken
    );
    const cached = this.analysisCache.get(cacheKey);
    if (cached) {
      return structuredClone(cached);
    }

    const snapshot = await this.buildSnapshot(
      request.wallet,
      request.limit,
      request.plaidAccessToken
    );
    this.analysisCache.set(cacheKey, snapshot.response);
    return structuredClone(snapshot.response);
  }

  async getCanonicalSnapshot(wallet: string): Promise<AnalysisSnapshot> {
    const snapshot = await this.buildSnapshot(wallet, DEFAULT_TRANSACTION_LIMIT);
    return {
      response: snapshot.response,
      totalSpendUsd: snapshot.totalSpendUsd,
      totalCo2eGrams: snapshot.totalCo2eGrams,
      categorySpendTotals: snapshot.categorySpendTotals,
      categoryEmissionTotals: snapshot.categoryEmissionTotals,
    };
  }

  clearWalletCache(wallet: string): void {
    for (const cacheKey of this.analysisCache.keys()) {
      if (cacheKey.startsWith(`${wallet}:`)) {
        this.analysisCache.delete(cacheKey);
      }
    }
  }
}

export const emissionsService = new EmissionsService();

File: app/services/emissions_service.py
"""Deterministic transaction analysis and emissions estimation."""

from __future__ import annotations

from decimal import Decimal

from contracts.python.contracts import (
    AnalyzeTransactionsRequest,
    AnalyzeTransactionsResponse,
    AnalyzedTransaction,
    DEFAULT_TRANSACTION_LIMIT,
    EmissionCategory,
)

from app.core.constants import (
    CATEGORY_ORDER,
    EMISSION_FACTORS,
    KEYWORD_CATEGORY_MAP,
    MCC_CATEGORY_MAP,
    REFERENCE_ANALYZED_AT,
)
from app.core.deterministic import decimal_to_float, round_decimal
from app.models.domain import AnalysisSnapshot, RawTransaction
from app.services.transaction_provider import transaction_provider
from app.state.store import user_store


class EmissionsService:
    def _get_raw_ledger(self, wallet: str, plaid_access_token: str | None = None) -> list[RawTransaction]:
        return user_store.get_or_create_ledger(
            wallet,
            lambda: transaction_provider.get_transactions(wallet, plaid_access_token),
        )

    def classify_transaction(self, raw_transaction: RawTransaction) -> EmissionCategory:
        if raw_transaction.mcc_code and raw_transaction.mcc_code in MCC_CATEGORY_MAP:
            return MCC_CATEGORY_MAP[raw_transaction.mcc_code]

        lowered = raw_transaction.description.lower()
        for keywords, category in KEYWORD_CATEGORY_MAP:
            if any(keyword in lowered for keyword in keywords):
                return category
        return EmissionCategory.OTHER

    def _build_snapshot(
        self,
        wallet: str,
        limit: int,
        plaid_access_token: str | None = None,
    ) -> AnalysisSnapshot:
        ledger = self._get_raw_ledger(wallet, plaid_access_token)
        selected = ledger[:limit]

        category_breakdown = {category: Decimal("0") for category in CATEGORY_ORDER}
        category_spend = {category: Decimal("0") for category in CATEGORY_ORDER}
        analyzed_transactions: list[AnalyzedTransaction] = []
        total_spend = Decimal("0")
        total_co2e = Decimal("0")

        for raw_transaction in selected:
            category = self.classify_transaction(raw_transaction)
            emission_factor = EMISSION_FACTORS[category]
            co2e_grams = round_decimal(raw_transaction.amount_usd * emission_factor, 2)

            analyzed_transactions.append(
                AnalyzedTransaction(
                    transaction_id=raw_transaction.transaction_id,
                    description=raw_transaction.description,
                    amount_usd=decimal_to_float(raw_transaction.amount_usd, 2),
                    mcc_code=raw_transaction.mcc_code,
                    category=category,
                    co2e_grams=decimal_to_float(co2e_grams, 2),
                    emission_factor=decimal_to_float(emission_factor, 2),
                    date=raw_transaction.date,
                )
            )

            total_spend += raw_transaction.amount_usd
            total_co2e += co2e_grams
            category_spend[category] += raw_transaction.amount_usd
            category_breakdown[category] += co2e_grams

        response = AnalyzeTransactionsResponse(
            wallet=wallet,
            transaction_count=len(analyzed_transactions),
            total_co2e_grams=decimal_to_float(total_co2e, 2),
            category_breakdown={category: decimal_to_float(category_breakdown[category], 2) for category in CATEGORY_ORDER},
            transactions=analyzed_transactions,
            analyzed_at=REFERENCE_ANALYZED_AT,
        )

        return AnalysisSnapshot(
            response=response,
            total_spend_usd=round_decimal(total_spend, 2),
            total_co2e_grams=round_decimal(total_co2e, 2),
            category_spend_totals={category: round_decimal(category_spend[category], 2) for category in CATEGORY_ORDER},
            category_emission_totals={category: round_decimal(category_breakdown[category], 2) for category in CATEGORY_ORDER},
        )

    def analyze_transactions(self, request: AnalyzeTransactionsRequest) -> AnalyzeTransactionsResponse:
        cached = user_store.get_cached_analysis(request.wallet, request.limit)
        if cached:
            return cached

        snapshot = self._build_snapshot(
            wallet=request.wallet,
            limit=request.limit,
            plaid_access_token=request.plaid_access_token,
        )
        user_store.set_cached_analysis(request.wallet, request.limit, snapshot.response)
        return snapshot.response

    def get_canonical_snapshot(self, wallet: str) -> AnalysisSnapshot:
        return self._build_snapshot(wallet=wallet, limit=DEFAULT_TRANSACTION_LIMIT)


emissions_service = EmissionsService()
